### PR TITLE
Select time zone for interpreting `TIMESTAMP WITHOUT TIME ZONE`

### LINKED ISSUE #2: Date only are parsed in UTC timezone, should be local time
commit 47ba3b0cf89814255c31549ae840191f8be3f0df breaks the date only parser.

According to [this link](http://dygraphs.com/date-formats.html), `2011-10-10T14:48:00` are interpreted in the UTC timezone, while `2011-10-10 14:48:00` is in the local time zone -the expected behavior, at least the one given in the comment:

```javascript
  // Force YYYY-MM-DD dates to be parsed as local time
  return DATE.test(isoDate) ?
    new Date(isoDate + 'T00:00:00') :
    null
```

A better implementation may be to use the matches of the regex:

```javascript
  var matches = DATE.exec(isoDate)
  var year = parseInt(matches[1], 10)
  var month = parseInt(matches[2], 10) - 1
  var day = matches[3]
  // Force YYYY-MM-DD dates to be parsed as local time
  return matches ?
    new Date(year, month, day) :
    null
```

### PR DESCRIPTION
**Pull request by dmurvihill**
This is a draft proposal to fix #2 and allow `postgres-date` to return correct Date values when the server and client are not in the same time zone.

## New parameter `serverTz`
Besides the unzoned timestamp itself, users who know the server time zone could pass a time zone name or numeric offset to `parseDate`, and get a proper Date object in the client's time zone. Downstream clients such as `node-postgres` could then implement tooling to fetch the server's time zone and convert dates transparently.

Many time zones change their offset for Daylight Saving Time. To get an accurate offset requires an un-zoned timestamp and a time zone. I believe the offset calculation can be done with zero additional dependencies using the `Intl` API. If not, the [`Temporal` polyfill](https://www.npmjs.com/package/@js-temporal/polyfill) will do the job (at a hefty but temporary cost of 3.4 MB until the standard is merged and starts shipping with engines).

## Handling Ambiguous Timestamps
Some timestamps may occur twice, or not at all, in the server's time zone. How these timestamps are handled depends on the implementation. There is a second option to handle these like Javascript `Date()`, like Postgres casts from `timestamp` to `timestamptz`, or to reject them entirely (defaults to postgres).

## Example Calls
In these examples, the client is in UTC+1 year-round and does not observe Daylight Saving Time.
```javascript
// Most users
parseDate('2026-01-01 00:00:00', 'America/New_York'); // 2026-01-01T06:00:00+01:00

// Weirder situations
parseDate('2026-01-01 00:00:00', 360); // 2026-01-01T07:00:00+01:00
parseDate('2025-11-02 01:30:00', 'America/New_York') // '2025-11-02T07:30:00+01:00'
parseDate('2025-11-02 01:30:00', {
  serverTz: 'America/New_York',
  tzMode: 'javascript',
}); // '2025-11-02T06:30:00+01:00'
parseDate('2025-11-02 01:30:00', {
  serverTz: 'America/New_York',
  tzMode: 'strict',
}); // Error
```

### CODE REVIEW

**Review by bendrucker — COMMENTED** [review]
Seems reasonable, I can think through the edge cases in more detail. What do we need to do to take this forward from a draft? Any specific open questions you're looking to answer or is the code effectively ready for review as-is?

**Comment by dmurvihill** [comment]
I was most worried about whether you would accept a new dependency on `Temporal`.

With that solved, the biggest obstacle is that none of the production code has been implemented (only tests).

I would like to take another pass at the test suite and look for missing edge cases, and request that you do so as well. (for example, Postgres timestamps are valid farther into the future than JS Dates, but not as far into the past. So that will be fun.)

I'd also like to see if there's a clever way to get a union type of all the valid IANA time zone names, rather than using `string`.

Finally, now would be a good time to nitpick the API design.

**Comment by dmurvihill** [comment]
Starting on implementation.

For local offset disambiguation, I am going to support `earlier` and `later` in addition to `postgres` and `javascript`. `strict` will be renamed to `reject`. This is for consistency with [Temporal.ZonedDateTime](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Temporal/ZonedDateTime).

**Comment by dmurvihill** [comment]
Ready for review now.

Adds a parameter for parsing `TIMESTAMP WITHOUT TIME ZONE` strings in an
arbitrary time zone or numeric UTC offset. After the date and time
components have been parsed, if the time zone is not present in the
timestamp string and the user has specified a time zone, `postgres-date`
uses the new `Temporal` API to compute the offset in the user-specified
time zone at the given local time. Timestamps with ambiguous offsets the
same way as Postgres does when casting `timestamp` to `timestamptz`, but
the user can select different behavior (including JavaScript-compatible
behavior) through an interface similar to`Temporal.ZonedDateTime`.

The polyfill is 736 KB fully installed. Minimum requirements are:
- Chrome 60
- Firefox 55
- Safari 11.1
- Safari iOS 11.3
- Edge 79
- Node.js 14

This means the new API will not work on Node 12. However, the polyfill
is lazy-loaded; Node 12 users can simply continue to use the existing
API.

In these examples, the client is in UTC+1 year-round and does not
observe Daylight Saving Time.
```javascript
// Most users
parseDate('2026-01-01 00:00:00', 'America/New_York'); // 2026-01-01T06:00:00+01:00

// Weirder situations
parseDate('2026-01-01 00:00:00', 360); // 2026-01-01T07:00:00+01:00
parseDate('2025-11-02 01:30:00', 'America/New_York') // '2025-11-02T07:30:00+01:00'
parseDate('2025-11-02 01:30:00', {
  timeZone: 'America/New_York',
  disambiguation: 'javascript',
}); // '2025-11-02T06:30:00+01:00'
parseDate('2025-11-02 01:30:00', {
  timeZone: 'America/New_York',
  disambiguation: 'reject',
}); // Error
```

**Comment by bendrucker** [comment]
Thanks, I haven't had time to look at this, but it's on my list.

**Comment by bendrucker** [comment]
> I believe the offset calculation can be done with zero additional dependencies using the Intl API. If not, the [Temporal polyfill](https://www.npmjs.com/package/@js-temporal/polyfill) will do the job (at a hefty but temporary cost of 3.4 MB until the standard is merged and starts shipping with engines).

I kind of glossed over this thinking that it was polyfilling a recent API not a proposal. It's a stage 4 proposal, but have you seen any indication of a proposed Node version where it would land?

**Comment by dmurvihill** [comment]
It's actually stage 3, expecting stage 4 in Q1 of this year. It will be included in ES2026 if it reaches stage 4 before their cutoff in February. Otherwise, ES2027.

[Latest news](https://thenewstack.io/es2026-solves-javascript-headaches-with-dates-math-and-modules/).

Node.js [uses the implementation in V8](https://github.com/nodejs/node/issues/57127), so they should make it available soon now that Chromium has shipped. If I had to guess, I would say Node 25. Safari is mid-implementation and Edge also uses V8.

**Comment by bendrucker** [comment]
Sorry for the slow reviews, addressed the tsd issue by requiring 20+. Looking at this now.

File: index.js [hunk 1]
```diff
@@ -10,11 +10,36 @@ const CHAR_CODE_Z = 'Z'.charCodeAt(0)
...
 class PGDateParser {
-  constructor (dateString) {
+  constructor (dateString, options) {
     this.dateString = dateString
     this.pos = 0
     this.stringLen = dateString.length
+
+    if (typeof options === 'object') {
+      this.timeZone = options.timeZone
+      if (options.disambiguation === 'undefined' || options.disambiguation === 'postgres') {
```
Comment (bendrucker) [inline index.js]: This compares against the string `'undefined'` rather than the value `undefined`, so the default won't apply when `disambiguation` is omitted.

File: index.js [hunk 2]
```diff
@@ -109,14 +134,37 @@ class PGDateParser {
...
+          loadTemporalPolyfill()
+        }
+        // eslint-disable-next-line no-undef
+        const pdt = Temporal.PlainDateTime.from({
+          year: date.year,
+          month: date.month + 1,
+          day: date.day,
+          hour: time.hours,
+          minute: time.minutes,
+          second: time.seconds,
```
Comment (bendrucker) [inline index.js:173]: `PlainDateTime` is timezone-agnostic — the `timeZone` property isn't part of its API and [may be rejected](https://tc39.es/proposal-temporal/docs/plaindatetime.html) by spec-compliant implementations. It's already used correctly in the `toZonedDateTime` call below so this line can just be removed.

File: package.json [hunk 3]
```diff
@@ -21,11 +21,14 @@
     "date",
     "parser"
   ],
-  "dependencies": {},
+  "dependencies": {
+    "temporal-polyfill": "^0.3.0"
```
Comment (bendrucker) [inline package.json]: Since the polyfill is only needed for the timezone feature and native Temporal is shipping in engines soon, could this be an optional peer dependency instead? Users who need timezone support install it explicitly, and once native Temporal lands they can drop it. The lazy `require` pattern already supports this — it'd just need a better error when the polyfill isn't installed.

File: index.js [hunk 4]
```diff
@@ -10,11 +10,36 @@ const CHAR_CODE_Z = 'Z'.charCodeAt(0)
 const CHAR_CODE_MINUS = '-'.charCodeAt(0)
 const CHAR_CODE_PLUS = '+'.charCodeAt(0)
 
+let temporalLoaded = false
+
+function loadTemporalPolyfill () {
+  temporalLoaded = true
+  if (globalThis.Temporal === undefined) {
+    globalThis.Temporal = require('temporal-polyfill').Temporal
+  }
```
Comment (bendrucker) [inline index.js]: Assigning to `globalThis` could conflict with other polyfills or native implementations that differ from the polyfill. This can be kept local:

```js
function loadTemporal () {
  return globalThis.Temporal ?? require('temporal-polyfill').Temporal
}
```

**Review by bendrucker — CHANGES_REQUESTED** [review]

**Comment by dmurvihill** [comment]
Rebased on `upstream/master`

File: index.js (hunk 1 above)
Comment (dmurvihill) [reply index.js]: Thanks, fixed.

File: index.js (hunk 2 above)
Comment (dmurvihill) [reply index.js:173]: Fixed

File: index.js (hunk 4 above)
Comment (dmurvihill) [reply index.js]: Yep, done.

File: package.json (hunk 3 above)
Comment (dmurvihill) [reply package.json]: `temporal-polyfill` is now an optional peer dep.

Comment (dmurvihill) [reply package.json]: Also added `@js-temporal/polyfill` as an optional peer dep. If the global `Temporal` is not available, it checks for both and loads the first available.

File: index.js [hunk 5]
```diff
@@ -10,11 +10,57 @@ const CHAR_CODE_Z = 'Z'.charCodeAt(0)
...
+
 class PGDateParser {
-  constructor (dateString) {
+  constructor (dateString, options) {
     this.dateString = dateString
     this.pos = 0
     this.stringLen = dateString.length
+
+    if (typeof options === 'object') {
+      this.timeZone = options.timeZone
```
Comment (bendrucker) [inline index.js]: When `options.timeZone` is a number (e.g. `parse(winter, { timeZone: 123 })`), it gets assigned to `this.timeZone` rather than `this.offset`. Then in `getTZOffset`, it's passed to `pdt.toZonedDateTime(this.timeZone, ...)` which expects an IANA string and will throw a `RangeError`.

The shorthand `parse(str, 123)` works because it hits the `typeof options === 'number'` branch. This branch needs to check `typeof options.timeZone` and route to `this.offset` for numbers:

```js
if (typeof options.timeZone === 'number') {
  this.offset = options.timeZone
} else {
  this.timeZone = options.timeZone
}
```

File: index.js [hunk 6]
```diff
@@ -10,11 +10,57 @@ const CHAR_CODE_Z = 'Z'.charCodeAt(0)
...
+}
+
 class PGDateParser {
-  constructor (dateString) {
+  constructor (dateString, options) {
     this.dateString = dateString
     this.pos = 0
     this.stringLen = dateString.length
+
+    if (typeof options === 'object') {
```
Comment (bendrucker) [inline index.js]: `typeof null === 'object'`, so `parse('2026-01-01', null)` enters this branch and crashes with `TypeError: Cannot read properties of null`. This is a plausible caller pattern when conditionally passing a timezone:

```js
const tz = config.timezone || null
parse(dateStr, tz)
```

Needs a null guard: `if (options !== null && typeof options === 'object')`

File: index.js [hunk 7]
```diff
@@ -10,11 +10,57 @@ const CHAR_CODE_Z = 'Z'.charCodeAt(0)
...
+  for (i = 0; i < paths.length && module === null; i++) {
+    try {
+      module = require(paths[i])
+    } catch (e) {
+      if (error === null) {
+        error = e.code === 'MODULE_NOT_FOUND'
+          ? new Error(MSG_MISSING_TEMPORAL, { cause: e })
+          : e
+      }
+    }
```
Comment (bendrucker) [inline index.js]: `error` is only assigned on the first failure. If the first polyfill is missing (`MODULE_NOT_FOUND`) but the second is installed and throws a real error (e.g. syntax error), the real error is discarded and the user sees the "install a polyfill" message even though a polyfill *is* installed.

Should prefer non-`MODULE_NOT_FOUND` errors:

```js
catch (e) {
  if (e.code !== 'MODULE_NOT_FOUND') {
    error = e
  } else if (error === null) {
    error = new Error(MSG_MISSING_TEMPORAL, { cause: e })
  }
}
```

File: index.js [hunk 8]
```diff
@@ -109,14 +155,33 @@ class PGDateParser {
...
     if (this.pos === this.stringLen - 3 && this.dateString.slice(this.pos, this.pos + 3) === '+00') {
       this.pos += 3
       return 0
     }
 
     if (this.stringLen === this.pos) {
+      if (this.offset !== undefined) {
+        return this.offset * 60 * 1000
+      } else if (this.timeZone !== undefined) {
+        // eslint-disable-next-line no-undef
```
Comment (bendrucker) [inline index.js]: `temporal` is a locally defined function — this suppression shouldn't be needed.

**Review by bendrucker — CHANGES_REQUESTED** [review]
Thanks for the revisions, if you could also extract polyfill-related things to a `temporal.js` module that would help simplify its eventual removal.

File: index.js (hunk 5 above)
Comment (dmurvihill) [reply index.js]: I adjusted the API in response to this. Now, the `options` object can have either a numeric `offset` or a string `timeZone`, but not both; the mutual exclusivity and types of each option are enforced.

File: index.js (hunk 6 above)
Comment (dmurvihill) [reply index.js]: Fixed with `if (options === null) { options = undefined; }` at the top. For TypeScript users, the type definitions still prohibit this pattern.

File: index.js (hunk 7 above)
Comment (dmurvihill) [reply index.js]: Good thinking, fixed.

File: index.js (hunk 8 above)
Comment (dmurvihill) [reply index.js]: Fixed

**Comment by dmurvihill** [comment]
Changes since yesterday:
- Polyfill loading moved to `temporal.js` (and corresponding tests moved as well)
- Fixed precedence of errors in Polyfill loading
- Split `options.timeZone` into two exclusive options `offset` and `timeZone`, and improved type validation
- Added `options.temporal` to allow passing a custom `Temporal` implementation

File: package.json
Comment (bendrucker) [inline package.json:1]: `temporal.js` is missing from the `files` array. It's `require`'d at runtime from `index.js` but won't be included in the npm package. Anyone using named timezone support will get `MODULE_NOT_FOUND` for `./temporal`.

File: index.d.ts [hunk 9]
```diff
@@ -1,3 +1,19 @@
+type MaybeTemporal = typeof globalThis extends { Temporal: infer T } ? T : unknown;
+
+interface ParseDateBaseOptions {
+  disambiguation?: 'earlier' | 'later' | 'reject' | 'postgres' | 'javascript';
+  temporal?: MaybeTemporal;
+}
+
+type ParseDateTimeZoneOptions = ParseDateBaseOptions & { timeZone?: string };
+type ParseDateOffsetOptions = ParseDateBaseOptions & { offset?: number };
```
Comment (bendrucker) [inline index.d.ts]: Two type issues:

1. **Mutual exclusivity not enforced.** Since both `timeZone` and `offset` are optional, `{ timeZone: 'UTC', offset: 300 }` compiles when passed indirectly via a `ParseDateOptions` variable. Use the `never` pattern:

```ts
type ParseDateTimeZoneOptions = ParseDateBaseOptions & { timeZone?: string; offset?: never };
type ParseDateOffsetOptions = ParseDateBaseOptions & { offset?: number; timeZone?: never };
```

2. **`MaybeTemporal` fallback should be `never`, not `unknown`.** When Temporal types aren't available, `temporal?: unknown` accepts any value. `never` makes the property non-settable, which is the correct constraint.

File: index.js [hunk 10]
```diff
@@ -11,10 +11,43 @@ const CHAR_CODE_MINUS = '-'.charCodeAt(0)
...
+        this.disambiguation = 'compatible'
+      } else {
+        this.disambiguation = options.disambiguation
+      }
+    } else if (typeof options === 'number') {
+      this.offset = options
+    } else {
+      this.timeZone = options
+      this.disambiguation = 'later' // Consistent with Postgres
+    }
```
Comment (bendrucker) [inline index.js:60]: This branch catches `undefined`, `string`, `boolean`, `bigint`, `symbol`, and `function`. Passing `true`, a `Symbol`, or a function silently stores it as `this.timeZone` and only fails later (or never, if the string has an embedded offset).

Should guard on `typeof options === 'string'` and reject unexpected types:

```js
} else if (typeof options === 'string') {
  this.timeZone = options
  this.disambiguation = 'later'
} else if (options !== undefined) {
  throw new TypeError(`Expected string, number, or options object, got ${typeof options}`)
}
```

File: index.js [hunk 11]
```diff
@@ -11,10 +11,43 @@ const CHAR_CODE_MINUS = '-'.charCodeAt(0)
...
+      } else if (options.offset !== undefined) {
+        throw new TypeError(`Expected numeric offset in minutes ahead of UTC, got offset=${options.offset.toString()}`)
+      }
+      if (options.disambiguation === undefined || options.disambiguation === 'postgres') {
+        this.disambiguation = 'later'
+      } else if (options.disambiguation === 'javascript') {
+        this.disambiguation = 'compatible'
+      } else {
+        this.disambiguation = options.disambiguation
+      }
```
Comment (bendrucker) [inline index.js]: The `else` passes `disambiguation` through verbatim without validation. Typos like `'rject'` are silently stored and only fail later in `toZonedDateTime` with a cryptic Temporal `RangeError`. With offset-only usage, the invalid value is never exercised at all.

Also, `disambiguation` is only meaningful with `timeZone` — numeric offsets have no DST ambiguity. `{ offset: -300, disambiguation: 'reject' }` silently ignores it.

Validate against the known set:

```js
const VALID_DISAMBIGUATIONS = new Set(['postgres', 'javascript', 'earlier', 'later', 'reject'])

if (options.disambiguation !== undefined && !VALID_DISAMBIGUATIONS.has(options.disambiguation)) {
  throw new TypeError(`Invalid disambiguation '${options.disambiguation}'. Expected one of: ${[...VALID_DISAMBIGUATIONS].join(', ')}`)
}
```

File: index.js [hunk 12]
```diff
@@ -109,14 +142,33 @@ class PGDateParser {
...
+        const pdt = Temporal.PlainDateTime.from({
+          year: date.year,
+          month: date.month + 1,
+          day: date.day,
+          hour: time.hours,
+          minute: time.minutes,
+          second: time.seconds,
+          millisecond: time.milliseconds
+        })
+        const utc = pdt.toZonedDateTime('UTC').toInstant()
```
Comment (bendrucker) [inline index.js]: Invalid IANA timezone names (e.g. `'America/NewYork'`) produce a raw Temporal `RangeError` with no context. This is inconsistent with the parser's convention of returning `null` for unparseable inputs — callers checking for `null` will get an unexpected exception.

Consider wrapping with context:

```js
try {
  const zdt = pdt.toZonedDateTime(this.timeZone, {
    disambiguation: this.disambiguation
  }).toInstant()
} catch (e) {
  if (e instanceof RangeError) {
    throw new RangeError(
      `Invalid time zone: '${this.timeZone}'`,
      { cause: e }
    )
  }
  throw e
}
```

File: index.js [hunk 13]
```diff
@@ -11,10 +11,43 @@ const CHAR_CODE_MINUS = '-'.charCodeAt(0)
...
+
+    if (typeof options === 'object') {
+      this.temporal = options.temporal
+      if (options.timeZone !== undefined && options.offset !== undefined) {
+        throw new TypeError("'offset' cannot be combined with 'timeZone'")
+      }
+      if (typeof options.timeZone === 'string') {
+        this.timeZone = options.timeZone
+      } else if (options.timeZone !== undefined) {
+        throw new TypeError(`Expected string time zone name, got timeZone=${options.timeZone.toString()}`)
```
Comment (bendrucker) [inline index.js]: `.toString()` on untrusted input can itself throw (e.g. for `Symbol` values), replacing the intended validation error with a confusing `TypeError: Cannot convert a Symbol value to a string`. Same issue on line 36 for `options.offset`.

Use `typeof` instead:

```js
throw new TypeError(`Expected string time zone name, got timeZone of type ${typeof options.timeZone}`)
```

File: temporal.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    try {
+      module = require(paths[i])
+    } catch (e) {
+      // Error precedence:
+      // 1. First error other than MODULE_NOT_FOUND
+      // 2. MODULE_NOT_FOUND for most preferred polyfill
+      if (e.code === 'MODULE_NOT_FOUND') {
+        error = error ?? new Error(MSG_MISSING_TEMPORAL, { cause: e })
+      } else {
+        error = error?.message === MSG_MISSING_TEMPORAL ? e : (error ?? e)
```
Comment (bendrucker) [inline temporal.js:24]: If a polyfill is installed but has a missing transitive dependency, `e.code` is still `MODULE_NOT_FOUND`. This branch replaces the real error with the generic "install a polyfill" message, even though the polyfill *is* installed.

Check whether the missing module is the polyfill itself:

```js
if (e.code === 'MODULE_NOT_FOUND' && e.message.includes(`'${paths[i]}'`)) {
  error = error ?? new Error(MSG_MISSING_TEMPORAL, { cause: e })
} else {
  error = error?.message === MSG_MISSING_TEMPORAL ? e : (error ?? e)
}
```

File: index.js [hunk 15]
```diff
@@ -109,14 +142,33 @@ class PGDateParser {
     return null
   }
 
-  getTZOffset () {
+  getTZOffset (date, time) {
     // special handling for '+00' at the end of  - UTC
     if (this.pos === this.stringLen - 3 && this.dateString.slice(this.pos, this.pos + 3) === '+00') {
       this.pos += 3
       return 0
     }
```
Comment (bendrucker) [inline index.js:161]: When the date string contains an explicit timezone suffix (e.g. `'2026-01-01T00:00:00+05:00'`), parsing hits the offset branches above before reaching here. The user-provided `timeZone`/`offset` option is silently ignored.

This is reasonable behavior (the string's embedded timezone should take precedence), but it should be documented in the README and tested explicitly. Users configuring a server timezone may not realize it's bypassed for `TIMESTAMPTZ` values.

File: test-temporal.js [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+          () => withFailingRequire(),
+          e,
+          'Prefer to show errors related to polyfills that are installed but failed to load'
+        )
+      } finally {
+        restoreModuleLoader()
+      }
+    }
+
+    globalThis.Temporal = Temporal
```
Comment (bendrucker) [inline test-temporal.js:95]: `Temporal` is not imported or declared in this file. `globalThis.Temporal` was set to `undefined` on line 20, so this assigns `undefined`. The "Prefers global Temporal over polyfill" assertion below is not actually testing that behavior.

Should be `globalThis.Temporal = oldTemporal` or an explicit `require('temporal-polyfill').Temporal`.

File: test-temporal.js [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    } finally {
+      restoreModuleLoader()
+    }
+
+    if (supportedPolyfills.length > 1) {
+      const restoreModuleLoader = throwErrorOnLoad([supportedPolyfills[1]], e)
+      try {
+        for (let i = 0; i < supportedPolyfills.length; i++) {
+          if (i !== 1) {
+            stubs[i] = null
```
Comment (bendrucker) [inline test-temporal.js]: `stubs[i]` uses numeric index as key instead of the polyfill name. Should be `stubs[supportedPolyfills[i]]`.

**Review by bendrucker — CHANGES_REQUESTED** [review]
Thanks for keeping after this, getting close!

File: package.json
Comment (dmurvihill) [reply package.json:1]: Any suggestions how to test this?

File: index.js (hunk 13 above)
Comment (dmurvihill) [reply index.js]: Are you sure?

```
dolan@Dolans-MacBook-Pro postgres-date % node
Welcome to Node.js v24.11.0.
Type ".help" for more information.
> Symbol('test').toString()
'Symbol(test)'
> new Symbol('test').toString()
Uncaught TypeError: Symbol is not a constructor
    at new Symbol (<anonymous>)
> 
dolan@Dolans-MacBook-Pro postgres-date % nvm use 20
Now using node v20.18.0 (npm v10.8.2)
dolan@Dolans-MacBook-Pro postgres-date % node
Welcome to Node.js v20.18.0.
Type ".help" for more information.
> Symbol('test').toString()
'Symbol(test)'
> 
dolan@Dolans-MacBook-Pro postgres-date % nvm use 12
Now using node v12.22.12 (npm v6.14.16)
dolan@Dolans-MacBook-Pro postgres-date % node      
Welcome to Node.js v12.22.12.
Type ".help" for more information.
> Symbol('test').toString()
'Symbol(test)'
```

Seeing similar in FF 147 and Chromium 124.

Implicit string coercion on `Symbol` values can throw `TypeError`, because it follows a different code path. But my understanding is `Symbol` has its own `.toString()` implementation that is fully defined.

File: index.d.ts (hunk 9 above)
Comment (dmurvihill) [reply index.d.ts]: Mutual exclusivity between `timeZone` and `offset` is correctly enforced and tested now. `disambiguation` is now exclusive with `offset` too.

As for the `temporal` option, its purpose is to enable passing an arbitrary implementation when the global `Temporal` is unavailable (and presumably its types are unavailable as well). Falling back to `never` would defeat the purpose. Obviously, this puts the caller in a place where there be dragons. I could add a warning to the README?

File: index.js (hunk 10 above)
Comment (dmurvihill) [reply index.js:60]: Tightened type enforcement of `options` object.

File: index.js (hunk 11 above)
Comment (dmurvihill) [reply index.js]: `disambiguation` now enforces one of five enumerated values. `null` is not allowed. Also, `disambiguation` is now mutually exclusive of `offset`.

File: temporal.js (hunk 14 above)
Comment (dmurvihill) [reply temporal.js:24]: Fixed.

File: index.js (hunk 15 above)
Comment (dmurvihill) [reply index.js:161]: Added the requested tests and documentation.

File: test-temporal.js (hunk 16 above)
Comment (dmurvihill) [reply test-temporal.js:95]: Whups! Fixed

File: test-temporal.js (hunk 17 above)
Comment (dmurvihill) [reply test-temporal.js]: Fixed

File: package.json
Comment (dmurvihill) [reply package.json:1]: Added `temporal.js` to the `files` array, but still interested in adding test coverage for this case

**Comment by dmurvihill** [comment]
API change: `disambiguation` is now mutually exclusive of `offset`

**Comment by dmurvihill** [comment]
Exported `VALID_DISAMBIGUATIONS`. This array will be very useful for fuzzers.

**Comment by charmander** [comment]
From the pg side, this seems like a lot of added mandatory complexity for something that is usually wrong (`timestamptz` being the type that represents instants in time) and that could be a separate package – not that there’s any issue, I guess, with the “separate package” just being a different version of postgres-date, with pg continuing to depend on the old version for `timestamptz`.

At the same time, if Temporal is already in play, why wouldn’t `Temporal.PlainDateTime` itself be the most natural default parse for `timestamp without time zone`?

**Comment by bendrucker** [comment]
> that could be a separate package

Yeah this is probably a good idea, could keep it in this repo as a separate workspace and then publish independently

**Comment by dmurvihill** [comment]
Maybe it's best to try for consensus on these points before merging any change here. I'd rather contribute something pg-core is 100% comfortable with.

It's fair to say parsing a plain `TIMESTAMP` to an instant in time is "usually wrong" conceptually. But that's what `postgres-date` does and it's too late to change now. I agree, `PlainDateTime` is technically the correct return type, but switching `postgres-date` to return `Temporal` types seems like a mistake to me, and returning a mishmash of `Temporal` and `Date` types seems even worse.

There are many, many users who are not experts in Postgres abusing `TIMESTAMP` this way. For example, TypeORM implements [`CreateDateColumn`](https://typeorm.io/docs/help/decorator-reference/#createdatecolumn), [`UpdateDateColumn`](https://typeorm.io/docs/help/decorator-reference/#updatedatecolumn), and [`DeleteDateColumn`](https://typeorm.io/docs/help/decorator-reference/#deletedatecolumn) with `TIMESTAMP` by default and uses `pg` to serialize these columns back to `Date`, ensuring cut-rate agencies across the world will display incorrect record creation times for the forseeable future.

If your company is in this situation, a database migration from `TIMESTAMP` to `TIMESTAMPTZ` is the best solution, but it's not always practical and almost always takes a while to organize. Conversely, updating your app to add a time zone parameter to your `pg` client can roll out relatively quickly.

There is admittedly quite a bit of new logic here, but I've taken pains to keep it contained to a very small number of branches:
- Constructor `if (options === undefined)`
- `getTZOffset` `if (this.offset !== undefined)` / `else if (this.timeZone !== undefined)`, where those two variables are only set in the constructor when processing the `options`.

I'm reasonably sure that unless the second argument is set to some value other than `null` or `undefined`, the parser will behave exactly as before and nothing new will be loaded. In that sense, I wouldn't characterize the added complexity as "mandatory".

**Comment by charmander** [comment]
> But that's what `postgres-date` does and it's too late to change now.

I don’t see why it would be too late to change (in a new major version). It’s already been done for the SQL `date` type.

> Conversely, updating your app to add a time zone parameter to your `pg` client can roll out relatively quickly.

Sure, but adding a type override is almost as easy and avoids introducing new configuration. (Type override presets can even be wrapped in a package.) I think the out-of-the-box behaviour should be the correct and simple version.

**Comment by dmurvihill** [comment]
I don't want to minimize your concerns; I very much do see your point and don't have any quarrel with using a type override rather than introducing a new configuration parameter.

That said, the out-of-the-box behavior of `postgres-date` and `pg`, while simpler, is not correct at all. By default, `pg` assigns the JavaScript engine's local time zone to `timestamp` columns regardless of from what zone the timestamp values originated or whether they are meant to refer to a single instant at all. This is not an abstraction, it is causing heartache downstream.

Brian is [very firmly against](https://github.com/brianc/node-postgres/issues/993#issuecomment-3898789526) altering `pg`'s default handling of `TIMESTAMP WITHOUT TIME ZONE` even in a new major version (and I agree with his reasons). To me, it doesn't make sense to change `postgres-date`'s default behavior only for `pg` to then change the result back to preserve compatibility.

We're not going to get correct behavior by default at this point (unless the `pg` maintainer changes his mind, which would be fine with me). That leaves two suggested approaches: an additional configuration parameter to `pg` (in several places), or a separate `postgres-timestamp` package that can be configured as a type override. I'm happy with either one, as long as we get an easy way to opt in to the correct behavior.

**Comment by charmander** [comment]
Oh, yes, agreed. To be clear, what I’m advocating for is removing the `timestamp without time zone` type parser or replacing it with a `PlainDateTime`-like type, not keeping the existing broken behaviour – like was done with `date`. I know @brianc mentioned not changing the existing time zone, but I’m not sure if that also rules out switching to something entirely non-`Date`.

**Comment by dmurvihill** [comment]
Roger that. It might work to return a POJO:

```
{
  calendar: 'iso8601',
  year: 2021,
  month: 7,
  day: 1,
  hour: 12,
  minute: 34,
  second: 56,
  millisecond: 23,
}
```

If the client has access to `Temporal`, they could pass this directly into [`PlainDateTime.from()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/UTC) and be on their merry way. If not, they can de-structure it into [`Date.UTC`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/UTC) and perform epoch shifting manually.

I don't know if this would make the breaking change any more visible to linters or type checkers, though. I assume at the bottom level, pg is just parsing the text and doing a type assertion on the result. ORMs _might_ be able to type check it properly, which would be a win.

**Comment by dmurvihill** [comment]
After marinating, I think this change should merge into `postgres-date` as-is. `postgres-date`'s scope is:
> This package parses [date/time outputs](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-OUTPUT) from Postgres into Javascript Date objects. Its goal is to match Postgres behavior and preserve data accuracy.

Returning a type other than `Date` would be inappropriate for this package. Removing support for `timestamp without time zone` kind of makes sense, but it would be disruptive downstream and probably result in the same code being added to `pg` anyway, for backward compatibility.

The only way to match Postgres behavior and preserve data accuracy for `timestamp without time zone` (that I can think of) is to accept external time zone information and do complicated date/time logic (i.e. use `Temporal`). I don't see a way around it.

`pg` may never adopt the new functionality, and that's ok. It will still be available to `pg` users through the type override mechanism. We've done our best above to minimize the blast.

There _should_ be a new package, though. it should be called `postgres-temporal` and return `Temporal` types (if available) or POJOs with properties `year`, `month`, `day`, etc. (if not). I bet it could mostly be copied from the current version of `postgres-date`.

**Comment by bendrucker** [comment]
Sorry for the late reply, haven't had a chance to look at this recently. Feel free to publish under that name. And send a PR adding a link to the README here. 

We can continue on this. I'm not fully caught up on the discussion/background yet, but that way you can move forward with the temporal package.

**Comment by bendrucker** [comment]
Did you ever end up publishing the temporal package? Realizing this got stuck and want to figure out whether this is still something worth merging or worthy of a new package.

**Comment by dmurvihill** [comment]
Hi Ben,

It's on my list but I haven't got round to it. Of course, feel free to publish it yourself if you feel like. It's mostly your code after all.

Meantime, my view of this PR is still the same.

Cheers!
